examples/apps/tree-comparison/src/model/legacyTreeInventoryList.ts

```diff
@@ -0,0 +1,244 @@
+			LegacySharedTree.getFactory().type,
+		) as LegacySharedTree;
+
+		const inventoryNode: BuildNode = {
+			definition: "inventory",
+			traits: {
+				// REV: If I try to remove these sample items (leaving inventoryItems as an empty array)
+				// the trait is removed entirely (undefined).  It seems that empty traits are discarded?
+				// This is a problem when later I want to iterate over it (in hasInitialized, when
+				// building up my initial set of InventoryItem objects), since I'll get a not-iterable error.
+				inventoryItems: [
+					{
+						definition: "inventoryItem",
+						traits: {
+							// REV: I tried adding an explicit "ID" trait here and using that for tracking,
+							// rather than the NodeId itself.  However, when deleting a node the "removed" member
+							// of the "viewChange" event args only includes the NodeId and the nodes are already
+							// gone (cannot be retrieved).  So the ID is lost at that point.  I could maintain a
+							// separate NodeId -> ID mapping to track, but this feels kind of clunky and defeats
+							// the point of getting away from direct usage of NodeIds.
+							name: {
+								definition: "name",
+								payload: "nut",
+							},
+							quantity: {
+								definition: "quantity",
+								payload: 0,
+							},
+						},
+					},
+					{
+						definition: "inventoryItem",
+						traits: {
+							name: {
+								definition: "name",
+								payload: "bolt",
+							},
+							quantity: {
+								definition: "quantity",
+								payload: 0,
+							},
+						},
+					},
+				],
+			},
+		};
+		legacySharedTree.applyEdit(
+			Change.insertTree(
+				inventoryNode,
+				StablePlace.atStartOf({
+					parent: legacySharedTree.currentView.root,
+					label: "inventory" as TraitLabel,
+				}),
+			),
+		);
+
+		this.root.set(legacySharedTreeKey, legacySharedTree.handle);
+	}
+
+	/**
+	 * hasInitialized is run by each client as they load the DataObject.  Here we use it to set up usage of the
+	 * DataObject, by registering an event listener for changes to the inventory list.
+	 */
+	protected async hasInitialized() {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		this._tree = await this.root
+			.get<IFluidHandle<LegacySharedTree>>(legacySharedTreeKey)!
+			.get();
+
+		// REV: Not exactly sure why a checkout is required to get "viewChange" events, seems like it should be able
+		// to fire off the legacy shared tree itself?
+		const checkout = new EagerCheckout(this._tree);
+		// This event handler fires for any change to the tree, so it needs to handle all possibilities (change, add, remove).
+		checkout.on("viewChange", (before: TreeView, after: TreeView) => {
+			const { changed, added, removed } = before.delta(after);
+			for (const quantityNodeId of changed) {
+				const quantityNode = this.tree.currentView.getViewNode(quantityNodeId);
+				// REV: This is annoying to iterate over since I can't filter until I've retrieved the node object.
```

Comment:
seems like a bit of a misplaced gripe IMO :). Getting node ids here means you can choose to view the node in the 'before' view OR the 'after' view relatively easily. I think the more annoying thing here is:

```
// This event handler fires for any change to the tree, so it needs to handle all possibilities (change, add, remove).
```

which tends to lead to pretty hard to understand inval/code (even for your simple app here, the inval logic here is not straightforward to read and I'm sure it was not obvious to write either)

Reply:
Well, for example `.delta()` could return something more like:
```typescript
interface NodeCentricDelta {
    added: TreeViewNode[];
    removed: TreeViewNode[];
    changed: { before: TreeViewNode; after: TreeViewNode }[];
}
```

And then I could do nice stuff like
```typescript
const addedInventoryItemNodes = added.filter((node) => node.definition === "inventoryItem");
```

Other options that could help might include e.g. registering for specifically adds/removes to the "inventoryItems" trait, or only registering for changes to nodes of definition "quantity", etc. to filter pre-handler.  Feels like kind of a similar problem to DOM eventing - you can technically just register all your handlers on the root and deduce/filter to your intended target, but it's easier to register the handlers more locally.

The inval code is definitely a weak point, but most of it is just dancing around the node filtering and `nodeId->node->nodeId->node` dereferencing.  And ultimately I still cheated - adds/removes still cause the view to just re-retrieve the whole inventory list (didn't want to think through how to communicate the index of the added node).  I think React might smooth this over enough since I'm keeping the individual `InventoryItem` references stable, though I haven't taken the time to profile/confirm.  I do think this highlights why the other tree examples are built using a "whole view is invalid, re-retrieve everything" event rather than granular invalidation, but since one of the main selling points of the tree is handling large data it would be nice to do the right thing in an example.  TBH my inval logic for the new SharedTree is worse (it requires more deduction/assumptions about how the tree is mutating) :(

Updated comments to explain rather than complain :)